Mathlib/Order/FixedPoints.lean

```diff
@@ -0,0 +1,325 @@
+
+end Basic
+
+section Eqn
+
+variable [CompleteLattice α] [CompleteLattice β] (f : β →o α) (g : α →o β)
+
+-- Rolling rule
+theorem map_lfp_comp : f (lfp (g.comp f)) = lfp (f.comp g) :=
+  le_antisymm ((f.comp g).map_lfp ▸ f.mono (lfp_le_fixed _ <| congr_arg g (f.comp g).map_lfp)) <|
+    lfp_le _ (congr_arg f (g.comp f).map_lfp).le
+#align order_hom.map_lfp_comp OrderHom.map_lfp_comp
+
+theorem map_gfp_comp : f (gfp (g.comp f)) = gfp (f.comp g) :=
+  f.dual.map_lfp_comp (OrderHom.dual g)
+#align order_hom.map_gfp_comp OrderHom.map_gfp_comp
+
+-- Diagonal rule
+theorem lfp_lfp (h : α →o α →o α) : lfp (lfp.comp h) = lfp h.onDiag := by
+  let a := lfp (lfp.comp h)
+  refine' (lfp_le _ _).antisymm (lfp_le _ (Eq.le _))
+  · exact lfp_le _ h.onDiag.map_lfp.le
+  have ha : (lfp ∘ h) a = a := (lfp.comp h).map_lfp
+  calc
+    h a a = h a (lfp (h a)) := congr_arg (h a) ha.symm
+    _ = lfp (h a) := (h a).map_lfp
+    _ = a := ha
+#align order_hom.lfp_lfp OrderHom.lfp_lfp
+
+theorem gfp_gfp (h : α →o α →o α) : gfp (gfp.comp h) = gfp h.onDiag :=
+  @lfp_lfp αᵒᵈ _ <| (OrderHom.dualIso αᵒᵈ αᵒᵈ).symm.toOrderEmbedding.toOrderHom.comp
+    (OrderHom.dual h)
+#align order_hom.gfp_gfp OrderHom.gfp_gfp
+
+end Eqn
+
+section PrevNext
+
+variable [CompleteLattice α] (f : α →o α)
+
+theorem gfp_const_inf_le (x : α) : gfp (const α x ⊓ f) ≤ x :=
+  (gfp_le _) fun _ hb => hb.trans inf_le_left
+#align order_hom.gfp_const_inf_le OrderHom.gfp_const_inf_le
+
+/-- Previous fixed point of a monotone map. If `f` is a monotone self-map of a complete lattice and
+`x` is a point such that `f x ≤ x`, then `f.prev_fixed x hx` is the greatest fixed point of `f`
+that is less than or equal to `x`. -/
+def prevFixed (x : α) (hx : f x ≤ x) : fixedPoints f :=
+  ⟨gfp (const α x ⊓ f),
+    calc
+      f (gfp (const α x ⊓ f)) = x ⊓ f (gfp (const α x ⊓ f)) :=
+        Eq.symm <| inf_of_le_right <| (f.mono <| f.gfp_const_inf_le x).trans hx
+      _ = gfp (const α x ⊓ f) := (const α x ⊓ f).map_gfp
+      ⟩
+#align order_hom.prev_fixed OrderHom.prevFixed
+
+/-- Next fixed point of a monotone map. If `f` is a monotone self-map of a complete lattice and
+`x` is a point such that `x ≤ f x`, then `f.next_fixed x hx` is the least fixed point of `f`
+that is greater than or equal to `x`. -/
+def nextFixed (x : α) (hx : x ≤ f x) : fixedPoints f :=
+  { f.dual.prevFixed x hx with val := lfp (const α x ⊔ f) }
+#align order_hom.next_fixed OrderHom.nextFixed
+
+theorem prev_fixed_le {x : α} (hx : f x ≤ x) : ↑(f.prevFixed x hx) ≤ x :=
+  f.gfp_const_inf_le x
+#align order_hom.prev_fixed_le OrderHom.prev_fixed_le
+
+theorem le_next_fixed {x : α} (hx : x ≤ f x) : x ≤ f.nextFixed x hx :=
+  f.dual.prev_fixed_le hx
+#align order_hom.le_next_fixed OrderHom.le_next_fixed
+
+theorem next_fixed_le {x : α} (hx : x ≤ f x) {y : fixedPoints f} (h : x ≤ y) :
+    f.nextFixed x hx ≤ y :=
+  Subtype.coe_le_coe.1 <| lfp_le _ <| sup_le h y.2.le
+#align order_hom.next_fixed_le OrderHom.next_fixed_le
```

Comment:
```suggestion
theorem prevFixed_le {x : α} (hx : f x ≤ x) : ↑(f.prevFixed x hx) ≤ x :=
  f.gfp_const_inf_le x
#align order_hom.prev_fixed_le OrderHom.prevFixed_le

theorem le_nextFixed {x : α} (hx : x ≤ f x) : x ≤ f.nextFixed x hx :=
  f.dual.prev_fixed_le hx
#align order_hom.le_next_fixed OrderHom.le_nextFixed

theorem nextFixed_le {x : α} (hx : x ≤ f x) {y : fixedPoints f} (h : x ≤ y) :
    f.nextFixed x hx ≤ y :=
  Subtype.coe_le_coe.1 <| lfp_le _ <| sup_le h y.2.le
#align order_hom.next_fixed_le OrderHom.nextFixed_le
```
and the corresponding changes for the next 7 theorems (I know it is very tedious).